### PR TITLE
cpo: awsprivatelink: adopt existing Endpoints by tag

### PR DESCRIPTION
**Pull request by sjenning**
This should fix some CI issues we are seeing caused by the `AWSEndpointServiceReconciler` in the CPO operating on old `AWSEndpointService` status and creating more than one AWS Endpoint for a particular `AWSEndpointService` resource.

In the case the `status.endpointID` is not set, we query AWS for an Endpoint that matches tags on cluster InfraID and AWSEndpointService name.  If one is found, then we adopt it vs the old behavior of creating a new one, resulting in an orphaned AWS Endpoint.

Test procedure:
* create HC and wait for AWSEndpointServices to be fully created
* scale down HO
* scale down CPO
* edit AWSEndpointServices to remove finalizers
* delete AWSEndpointServices
* scale up HO (which will scale up CPO)
* AWSEndpointServices should be reconstructed with the exact same status as before

### CODE REVIEW

File: control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go [hunk 1]
```diff
@@ -112,6 +112,10 @@ func (r *PrivateServiceObserver) Reconcile(ctx context.Context, req ctrl.Request
 		},
 	}
 	if err := r.Get(ctx, client.ObjectKeyFromObject(svc), svc); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.log.Info("service not found")
```
Comment (ironcladlou) [inline control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go:116]: Given there's no prior context logged, what about logging which service wasn't found?

File: control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go [hunk 2]
```diff
@@ -337,28 +341,47 @@ func reconcileAWSEndpointService(ctx context.Context, awsEndpointService *hyperv
...
-				Tags:         apiTagToEC2Tag(hcp.Spec.Platform.AWS.ResourceTags),
-			}},
-		}
-		output, err := ec2Client.CreateVpcEndpointWithContext(ctx, input)
+
+		// Verify there is not already an Endpoint that we can adopt
+		// This can happen if we have a stale status on AWSEndpointService or encoutered
+		// an error updating the AWSEndpointService on the previous reconcile
+		output, err := ec2Client.DescribeVpcEndpointsWithContext(ctx, &ec2.DescribeVpcEndpointsInput{
+			Filters: apiTagToEC2Filter(awsEndpointService.Name, hcp.Spec.Platform.AWS.ResourceTags),
```
Comment (ironcladlou) [inline control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go:349]: Names are unique within a basedomain... is relying on name plus user-defined tags open to adoption bugs for clusters across domains?

Comment (sjenning) [reply control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go:349]: the infra id for the cluster must be unique and we are matching on a `{infra-id, service-name}` set

Comment (ironcladlou) [reply control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go:349]: gotcha, that works!

**Review by alvaroaleman — APPROVED** [review]
/hold

Looks good, holding so you can address Dan's comment

**Comment by ironcladlou** [comment]
/hold cancel

Can tweak logs in a followup, we need this merged ASAP

**Pull request merged**